skxray/correlation.py

```diff
@@ -0,0 +1,371 @@
+# IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE   #
+# POSSIBILITY OF SUCH DAMAGE.                                          #
+########################################################################
+
+"""
+
+This module is for functions specific to time correlation
+
+"""
+
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+import six
+import numpy as np
+import numpy.ma as ma
+import logging
+import time
+
+import skxray.core as core
+logger = logging.getLogger(__name__)
+
+
+def multi_tau_auto_corr(num_levels, num_bufs, labels, images):
+    """
+    This function computes one-time correlations.
+    
+    It uses a scheme to achieve long-time correlations inexpensively
+    by downsampling the data, iteratively combining successive frames.
+
+    The longest lag time computed is num_levels * num_bufs.
+
+    Parameters
+    ----------
+    num_levels : int
+        how many generations of downsampling to perform, i.e.,
+        the depth of the binomial tree of averaged frames
+
+    num_bufs : int, must be even
+        maximum lag step to compute in each generation of
+        downsampling
+
+    labels : array
+        labeled array of the same shape as the image stack;
+        each ROI is represented by a distinct label (i.e., integer)
+
+    images : iterable of 2D arrays
+        dimensions are: (rr, cc)
+
+    Returns
+    -------
+    g2 : array
+        matrix of one-time correlation
+        shape (num_levels, number of labels(ROI))
+
+    lag_steps : array
+        delay or lag steps for the multiple tau analysis
+        shape num_levels
+
+    Note
+    ----
+
+    This implementation is based on code in the language Yorick
+    by Mark Sutton, based on published work. [1]_
+
+    References
+    ----------
+
+    .. [1] D. Lumma, L. B. Lurio, S. G. J. Mochrie and M. Sutton,
+        "Area detector based photon correlation in the regime of
+        short data batches: Data reduction for dynamic x-ray
+        scattering," Rev. Sci. Instrum., vol 70, p 3274-3289, 2000.
+
+    """
+    # In order to calculate correlations for `num_bufs`, images must be
+    # kept for up to the maximum lag step. These are stored in the array
+    # buffer. This algorithm only keeps number of buffers and delays but
+    # several levels of delays number of levels are kept in buf. Each
+    # level has twice the delay times of the next lower one. To save
+    # needless copying, of cyclic storage of images in buf is used.
+
+    if num_bufs % 2 != 0:
+        raise ValueError("number of channels(number of buffers) in "
+                         "multiple-taus (must be even)")
+
+    if labels.shape == images[0].shape[1:]:
+        raise ValueError("Shape of an image should be equal to "
+                         "shape of the labels array")
+
+    # Which pixels are in each label?
+    labels, pixel_list = extract_label_indices(labels)
```

Comment:
can you rename the input `labels` to something like `label_mask`?  re-assigning a very different data structure to the same name is a code tick that drives me up the wall.